### ISSUE TITLE
.wrap of a method breaks callsame/nextsame/etc.

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2178]
## The Problem

Wrapping a method in a child class breaks MRO chaining for callsame/nextsame. Consider the following code:

```
     class Foo {
         has Int $.attr is rw;

         method m {
             say "<this is Foo:m>";
             callsame;
         }
     }

     class Bar is Foo {
         method m {
             say "<this is Bar::m>";
             callsame;
         }
     }

     class Baz is Bar {
         method m {
             say "<this is Baz::m>";
             callsame;
         }
     }

     Baz.^lookup( 'm' ).wrap( method { say "<this is WRAPPER:m>"; nextsame } );

     my $inst = Baz.new;

     $inst.m;
``` 

It produces the following output:

```
<this is WRAPPER:m>
<this is Baz::m>
```

Method `m` of class `Baz` cannot call the next method in MRO chain.

There is even worse case scenario if wrap is applied to `Bar::m` method. Replacing the wrapping line in the source with the following:

```
Bar.^lookup( 'm' ).wrap( method { say "<this is WRAPPER:m>"; nextsame } );
```

gives us:

```
<this is Baz::m>
Cannot invoke object with invocation handler in this context
```

## Environment
* Operating system: macOS 10.13
* Compiler version (`perl6 -v`): This is Rakudo version 2018.06 built on MoarVM version 2018.06

**Comment by vrurg** [comment]
There is interesting development for the situation when the method in `Bar` is wrapped. Adding a call to `nextcallee` into `Baz::m`:

```
     class Baz is Bar {
         method m {
             say "<this is Baz::m>";
             my &nc = nextcallee;
             #self.&nc;
             callsame;
         }
     }
```

avoid the exception about invocation handler. Instead the following output is produced:

```
<this is Baz::m>
<this is Foo:m>
```

It looks like this time `callsame` just got around `Bar::m` by simply ignoring it! By uncommenting the `self.&nc` call I get the expected output:

```
<this is Baz::m>
<this is WRAPPER:m>
<this is Bar::m>
<this is Foo:m>
```

but the point that it is produced by two separate calls. I.e., `callsame` from `Bar::m` did not pass the control to `Foo`, but from `Baz::m` did!

Now, the final trick: leave the `self.&nc` but comment out `callsame` in `Baz::m`. It raises another exception:

```
<this is Baz::m>
Cannot invoke this object (REPR: Uninstantiable; Callable)
  in method m at ./wrap.p6 line 23
  in block <unit> at ./wrap.p6 line 32
```

These tests were made with a recent rakudo checkout: `This is Rakudo version 2018.06-407-g3301ddb1c built on MoarVM version 2018.06-399-g4fc267b60`

**Comment by vrurg** [comment]
Closed via #2749
